### DISCUSSION TITLE
Suggestions to build open source projects

### DISCUSSION THREAD

**Discussion by AbcSxyZ** [ZigRazor/CXXGraph#91] [Ideas]
Hi,

I saw your post on [dev.to](https://dev.to/listings/collabs/looking-for-collaborators-contributors-for-open-source-project-31hf) where you're looking for collaborators. I explored your repository, I'm not a CXX user and even less doing ML learning, but I thought I may help you to attract more contributor by giving you some tips about Open Source project. I saw multiple area where improvement seem possible.

To be transparent, for now I do not have the experience of maintaining a successful OS project myself, but I'm studying the topic and hope to give you useful feedback and resources to think about how to attract more users & contributors. Actually, I'm starting a project called [Open Source Education](https://github.com/AbcSxyZ/Open-Source-Education) to educate about the topic.

First of all, I can warmly suggest you to take a look at *"Producing Open Source Software, How to Run a Successful Free Software Project"* by [Karl Fogel](https://producingoss.com/) & *"The cathedral and the bazaar"* of [Eric Raymond](http://www.catb.org/~esr/writings/cathedral-bazaar/cathedral-bazaar/).

To attract more contributor, I guess you should more target your users rather, Hacktoberfest contributor who do not use you repo won't stay here for mid/long term. It's about how you can make your actual users becoming potential contributors, and how to increase user base to have more potential people who will do this switch.

## Concrete improvement suggestion

As a fisrt tips & the most obvious to me, there is one major area of improvement : Documentation :)

Without speaking about your other repo having an empty doc, making the software difficult to use and understand, here you've a lot of content, it's more about the docs organization which can be more efficient.

Karl Fogel is speaking about `hacktivation energy`, the energy it cost to use and get involved in your software, as user and as contributor. From a maintainer perspective, your goal is to reduce the entry barrier as much as possible. For example here, how long it takes for the user to see the installation section & jump in with your software ?

As I see it, you can make it really more easy to install. The current installation is not at the top of the document (take a look some other repo when install instructions are coming), and you're even showing how to compile test before how to use your soft. Just by changing how your doc is structured may help to facilitate usage.

Without a deep thinking, I would see your README structured more like that: 
1. Your introduction sentence, before everything else (the table of content), let the user directly understand if he may or may not be interested in the soft
2. Your table of content
3. Install instructions
4. Link to documentation
5. List of algorithm

It may be a more intuitive path from a user point of view.

**I'm sure by improving your maintainer skill, you will be able to attract more people :) There is good resources to learn about it !**

**Comment by ZigRazor** [comment]
Hi @AbcSxyZ 
Thank you so much for this suggestions. You think you can help me to rework my readme in a more attractive way?
If you want help me you can open a PR with the Readme Reworked.
Thank you so much in advance,

**Reply by AbcSxyZ** [reply]
I've the ability to rework your README, but I won't do that, it's not my intent. It's a suggestion of improvement.

I prefer to help you to understanding some way to have a more usable & attractive open-source project. If I do all by myself and you don't understand why it's not really useful for you. I would love to see you having this kind of thought by yourself !

To help you to understand how to build an open-source project, to improve your maintainer skill. For here, but anywhere else :)
I will be interested in following the evolution, because I'm learning myself

**Reply by ZigRazor** [reply]
Ok, so only an intellectual collaboration. No more than suggestion, in any case I will try to apply your suggestions!
Thank you!

**Reply by AbcSxyZ** [reply]
Yes, I more interested in open source education, educate about how open source works, to contribute and maitain a project.

As you see, building an open source project is not only about releasing your source code. When you're the maintainer, you will use strategies to involve contributors in your project, and that's an other world ! I'm here to help you think about that and give some resources and feedback/discussion on the topic.

A lot of developers don't know how it works really, learning by experience. But there is plenty of content from people working on the subject which are really helpful to build a project :)

**Reply by ZigRazor** [reply]
What is the best site/platform in which I can advertise of my project?

**Reply by AbcSxyZ** [reply]
What kind of people are going to be interested in your software ? Depending of the answer, it can change :)

**Reply by ZigRazor** [reply]
Developers, and users interested in Graph argument

**Reply by AbcSxyZ** [reply]
You need to find people who will need your solution, to solve someone else problem. Someone might help you to contribute if he need it and find interest in co-producing the software with you. Otherwise, they probably won't care.

**Reply by AbcSxyZ** [reply]
Developers is too large. Also, do you know if there are similar projects to yours ? Your repo have something not already solved ?

A lot of time books are speaking about identifying related projects to avoid reinventing the wheel.

**Comment by ZigRazor** [comment]
If you want, I can invite you as a collaborator. You can do Experience with promotion of Open-Source-Project and help me to create a community! Let me know!

**Comment by ZigRazor** [comment]
@AbcSxyZ  Some other suggestion? 
Thank you so much !
